db/db_impl/db_impl_write.cc

```diff
@@ -512,15 +512,18 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
   }
   PERF_TIMER_START(write_pre_and_post_process_time);
 
+  if (!io_s.ok()) {
```

Comment:
I changed the error handling of `io_s` in this file since I was worried about the case when WriteToWAL returns corruption **and** `w.CallbackFailed()` is true: either there is an assert for `io_s.okay()` or there is no checking for `io_s` in this case before the change in this PR. I'm not familiar with the writer callback, whether the error handling change need to be included in this PR depends on if the above scenario is possible.

Reply:
Sorry, the longer I look at the existing code, the more confusing it becomes. It might be because `FinalStatus()` returns the non-callback failure first. However, if the callback failed, then the callback failure is the first failure and should be returned in `FinalStatus()`. So when `WriteToWAL()` and leader callback both failed, we should only record the callback failure. 

 Why we even proceed to `WriteToWAL()` after callback failure considering https://github.com/facebook/rocksdb/blob/ad135f3ffdf9a932000d289101867a43f101d5e4/db/write_callback.h#L18-L20 is a mystery to me. What happens to callback failures in non-leader writers is also unclear.

Reply:
Anyways, I don't want to derail this. This is setting a write group wide status when WriteToWAL() fails and that's good for me. Worst case looks like a `WriteToWAL()` error can be returned when the actual first failure was in the leader callback, which is fine with me.